kubeadm/v1.15.0/KubeClusterHelper.psm1

```diff
@@ -0,0 +1,1360 @@
+                try 
+                {
+                    m_serviceStatus.dwCurrentState = ServiceState.SERVICE_STOPPED;
+                    if (m_process != null)
+                    {
+                        m_process.Kill();
+                        m_process.WaitForExit();
+                        m_process.Close();
+                        m_process.Dispose();
+                        m_process = null;
+                    }
+                    Console.WriteLine("OnStop - Successfully stopped the service ");
+                } 
+                catch (Exception e)
+                {
+                    Console.WriteLine("OnStop - Failed to stop the service : " + e.Message);
+                    m_serviceStatus.dwCurrentState = ServiceState.SERVICE_RUNNING;
+                }
+                finally
+                {
+                    SetServiceStatus(ServiceHandle, ref m_serviceStatus);
+                    if (m_writer != null) m_writer.Flush();
+                }
+            }
+
+            public static void Main() {
+                System.ServiceProcess.ServiceBase.Run(new ScmService_$ServiceName());
+            }
+        }
+"@
+
+    Add-Type -TypeDefinition $svcSource -Language CSharp `
+        -OutputAssembly $SvcBinary -OutputType ConsoleApplication   `
+        -ReferencedAssemblies "System.ServiceProcess" -Debug:$false
+
+    return $SvcBinary
+}
+
+function RemoveService()
+{
+    param
+    (
+        [parameter(Mandatory=$true)] [string] $ServiceName
+    )
+    $src = Get-Service -Name $ServiceName  -ErrorAction SilentlyContinue
+    if ($src) {
+        Stop-Service $src
+        sc.exe delete $src;
+    }
+}
+
+function InstallDockerD()
+{
+    Param(
+    [ValidateSet("docker")] [parameter(Mandatory = $false)] $Version = "docker",
+    $DestinationPath
+    ) 
+    # Add path to this PowerShell session immediately
+    $env:path += ";$env:ProgramFiles\Docker"
+    # For persistent use after a reboot
+    $existingMachinePath = [Environment]::GetEnvironmentVariable("Path",[System.EnvironmentVariableTarget]::Machine)
+    [Environment]::SetEnvironmentVariable("Path", $existingMachinePath + ";$env:ProgramFiles\Docker", [EnvironmentVariableTarget]::Machine)
+
+    $cmd = get-command docker.exe -ErrorAction SilentlyContinue
+    if (!$cmd)
+    {
+        $dockerVersion = $Version
+        DownloadFile  "https://master.dockerproject.org/windows/x86_64/${dockerVersion}.zip" -Destination "$env:TEMP\$dockerVersion.zip" 
+        Expand-Archive -Path "$env:TEMP\$dockerVersion.zip" -DestinationPath $env:ProgramFiles -Force
+        dockerd --register-service
+        Start-Service Docker -ErrorAction Stop
+        $Global:Configuration += @{
+            InstallDocker = $true;
+        }
+        WriteKubeClusterConfig
+    }
+}
+
+function InstallDockerImages()
+{
+    if (!(docker images $Global:NanoserverImage -q))
+    {
+        docker pull $Global:NanoserverImage
+        if (!$LastExitCode) {
```

Comment:
This is failing for me, despite successful docker image pull. Should this be inverted to either
`if ($LastExitCode)` or `if (!($LastExitCode -eq 0))` ?

Reply:
Fixed by just checking docker images again